File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_Stoch_D_TwoAscending.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.model.indicators.StochJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.OK;
import static com.kishlaly.ta.model.BlockResultCodeJava.STOCH_D_NOT_ASCENDING_SCREEN_2;

/**
 * last %D stochastics above
 */
public class Long_ScreenTwo_Stoch_D_TwoAscending implements ScreenTwoBlock {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<StochJava> screen_2_Stochastic = (List<StochJava>) screen.indicators.get(IndicatorJava.STOCH);
        StochJava stoch2 = screen_2_Stochastic.get(screen_2_Stochastic.size() - 2);
        StochJava stoch1 = screen_2_Stochastic.get(screen_2_Stochastic.size() - 1);

        boolean ascendingStochastic = stoch2.getSlowD() < stoch1.getSlowD();
        if (!ascendingStochastic) {
            LogJava.recordCode(STOCH_D_NOT_ASCENDING_SCREEN_2, screen);
            LogJava.addDebugLine("Stochastic %D does not grow on the second screen");
            return new BlockResultJava(screen.getLastQuote(), STOCH_D_NOT_ASCENDING_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/stoch/Long_ScreenTwo_Stoch_D_TwoAscending.kt
package com.kishlaly.ta.analyze.tasks.blocks.two.stoch

import com.kishlaly.ta.analyze.tasks.blocks.two.ScreenTwoBlock
import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.Stochastic
import com.kishlaly.ta.utils.Log
import com.kishlaly.ta.utils.SymbolDataUtils

/**
 * last %D stochastics above
 */
class Long_ScreenTwo_Stoch_D_TwoAscending : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        val stoch = SymbolDataUtils(screen, Stochastic::class.java)
        val ascendingStochastic = stoch[2].slowD < stoch[1].slowD
        if (!ascendingStochastic) {
            Log.recordCode(BlockResultCode.STOCH_D_NOT_ASCENDING_SCREEN_2, screen)
            Log.addDebugLine("Stochastic %D does not grow on the second screen")
            return BlockResult(screen.lastQuote, BlockResultCode.STOCH_D_NOT_ASCENDING_SCREEN_2)
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}